### ISSUE TITLE
Requirements for salt

### ISSUE THREAD

**Issue by dreaming-augustin** [P-H-C/phc-winner-argon2#255]
I cannot find the proper requirements for the salt. I am confused by the official documentation and the C code examples.

The code sample of the front page of this project appear to suggest that we can pass a zero-initialized salt, probably implying that the salt will be properly generated internally:

  ```
    uint8_t salt[SALTLEN];
    memset( salt, 0x00, SALTLEN );

```     
       
On the other hand, the pdf documentation has only this to say about salt:

>   Primary inputs are message P and nonce S, which are password and salt, respectively, for the password hashing. (...)
> • Nonce S may have any length from 8 to 232 − 1 bytes (16 bytes is recommended for password hashing).
> (....)
> Recommended parameters: 5. Select the salt length. 128 bits is sufficient for all applications, but can be reduced to 64 bits in the case of space constraints.     
       
At this point, I am guessing but I am not sure that we do have to provide the salt ourselves, and provide our own implementation of a random salt. The length of the salt is discussed above, but I am still unsure of the accepted values for a salt: should it be alpha-numeric only? ASCII characters? 

A simple random salt implementation like the following:
```
#include <iostream>
#include <bsd/stdlib.h> // arc4random_buf

int main() {                                                                                                                                                                                  
        char myRandomSalt[50];
        arc4random_buf(myRandomSalt, sizeof myRandomSalt);
        std::cout << "myRandomSalt: " << myRandomSalt << std::endl;
        return 0;
}
```
outputs non-ascii, unprintable characters like:
`myRandomSalt: 7��bE�D�▒�$��=DD��В�F�H�Ӓ�N��C���h���b    �Q�9�
`   
     
Would that do?

thanks.

**Comment by sneves** [comment]
Per the specification, the salt is a sequence of `saltlen` bytes, with `saltlen` less than 2^32 - 1 and larger than 7. That is the only requirement on the salt, it can be anything.

To generate the salt, a random string of bytes like you have above is fine. It does not even have to be random; the only requirement is that the salt be unique, so a global counter would work as well. But random is usually easier to manage.

**Comment by dreaming-augustin** [comment]
Thank you sneves. I think I get it. 

I think many would accurately surmise that I am a begging and certainly out of my depth, here. 
The current documentation is certainly enough for highly experienced programmers. 
However, for programmers on the lower end of the spectrum, like myself, some parts are arcane and unclear. What's worse, there is a tendency by some irresponsible programmers to copy/paste code without understanding it.   It is obviously their responsibility, but maybe the code segment below, as found on this project's homepage, could be supplemented with some comments that would make clear how this library is meant to be used

```
   uint8_t salt[SALTLEN];
  memset( salt, 0x00, SALTLEN );   
```
So no, argon2 does not generate the salt itself, as a wrong interpretation of the above code snippet might suggest. 
It is also worth mentioning this as some libraries implement argon2 and also relieve the developer of providing a unique salt.     

The maintainers should feel free to close this issue, but I leave it open as an invitation to make relevant parts of the documentation more explicit.

**Comment by sneves** [comment]
That's fair.

**Comment by thekogans** [comment]
please excuse me hijacking your post, but since my question is salt related i figured i'd continue the discussion instead of starting a new thread.

firstly, i would like to thank the authors for making this distribution freely available to the rest of us. much appreciated.

i am having a hard time understanding the requirement for salt. is it or is it not optional. the code is actually quite schizophrenic on the subject. firstly, there's this code for core.c

```
    /* Validate salt (required param) */
    if (NULL == context->salt) {
        if (0 != context->saltlen) {
            return ARGON2_SALT_PTR_MISMATCH;
        }
    }

    if (ARGON2_MIN_SALT_LENGTH > context->saltlen) {
        return ARGON2_SALT_TOO_SHORT;
    }

    if (ARGON2_MAX_SALT_LENGTH < context->saltlen) {
        return ARGON2_SALT_TOO_LONG;
    }
```

why test for mismatch between pointer and length if the pointer is required to not be NULL. second, there's this code (also from core.c);

```
    if (context->salt != NULL) {
        blake2b_update(&BlakeHash, (const uint8_t *)context->salt,
                       context->saltlen);
    }
```

which seems to suggest that salt is optional. which is it?

**Comment by thekogans** [comment]
thank you, samuel for a very fast reply. as you were typing i came across the argon2 spec and, sure enough, salt is mandatory!

**Comment by nicraMarcin** [comment]
Hello, I have question about salt.
I salt is mandatory so, can it be one salt in whole application or I should store salt in user table where each user has different salt?

**Comment by SparkDustJoe** [comment]
It depends on the purpose for using Argon2.  For user logins, yes each salt should be different, and stored with the users' information.  You can apply a Secret or Additional Data in Argon2 that can be for an entire application, differentiating one application from another.  It should express a purpose, and be unique (maybe including a randomly generated, but constant once committed, integer).
"Application 2020-05-23 Login for Users 0F34C522" vs. "Application2 2020-04-13 Shared Key Generation 00EA598F"

**Comment by nicraMarcin** [comment]
Yes I'have found that I can provide additional secret string for whole application, but this isn't provided in high-level methods but should be
```
ARGON2_PUBLIC int argon2i_hash_encoded(const uint32_t t_cost,
                                       const uint32_t m_cost,
                                       const uint32_t parallelism,
                                       const void *pwd, const size_t pwdlen,
                                       const void *salt, const size_t saltlen,
                                       const size_t hashlen, char *encoded,
                                       const size_t encodedlen);
```
@SparkDustJoe 

> Application 2020-05-23 Login for Users 0F34C522 
shouldn't be a link?

**Comment by SparkDustJoe** [comment]
Most implementers don't have a need for the Secret field, but it is part of the spec, so if it isn't included in an overloaded function, that's worth requesting.

If the salt is included in the encoded string, then just store the whole string for that use (I don't remember if Argon2 natively stores/expresses that in the output or if this library does either, it's been a while since I've worked with the code). If the salt is not included in the output string, then it needs to be stored in a separate field in the same user row, either as raw binary in the database or as an encoded string itself (either Base32, Base64, Ascii85, or something along those lines). 

Not sure what you mean by "shouldn't be a link?" The name Application is just a placeholder for whatever you like, like "Github", "com.android.sudokuchampions", "DailyCrosswordStars" or whatever the application may be.  The integer could be replaced with the major version number of the application (care has to be taken here as users would have to update or "recover" their passwords if any input to the algorithm changes, as the resulting output would no longer match!). I was just giving suggestions based on what I see most recommended for Personalization Strings, static Secrets, or Additional Data fields in systems like this.

**Comment by nicraMarcin** [comment]
> Most implementers don't have a need for the Secret field, but it is part of the spec, so if it isn't included in an overloaded function, that's worth requesting.

Showed function argon2i_hash_encoded() is argon2 function from argon2.h, so I thik is worth to provide secret field to this function.


> If the salt is included in the encoded string, 

ok, this is vicious circle. For generate encoded string I have to provide salt string, but encoded string isn't generated yet so I cant't get salt from this string and salt and saltlength I have to provide in to argon2i_hash_encoded() function.


> Not sure what you mean by "shouldn't be a link?" 

I thought that you tried to point mi some details, inforamation, show some article so I thought `"Application 2020-05-23 Login for Users 0F34C522" vs. "Application2 2020-04-13 Shared Key Generation 00EA598F" should be links because google dosn't know about it :(
